modules/IRCChat.js

```diff
@@ -0,0 +1,138 @@
+function IRCChat(host, server_port, name, pass, channel, WelcomeMsg, isSecure, cmd_prefix) {
+	
+	var self = this;
+	var users = [];
+	var userRanks = {};
+	var topic = "";
+	var my_username = name;
+	
+	if (!host || !server_port || !name || !pass || !channel || !WelcomeMsg) { throw new Error("Missing argument! You need to specify host, port, name, channel and welcome message!"); }
+	if (!cmd_prefix) { cmd_prefix = "!"; }
+	if (typeof isSecure === "undefined") { isSecure = false; }
+	if (!channel.startsWith("#")) { channel = "#" + channel; }
+	
+	
+	if (typeof IRCConnections[host + ":" + server_port + "_" + name] === "undefined") {
+		IRCConnections[host + ":" + server_port + "_" + name] = new irc.Client(host, name, {
+			port: server_port,
+			userName: "NewXKitBot",
+			realName: "New-XKit Bot",
+			secure: isSecure,
+			autoRejoin: true
+		}).on('registered', function(message) {
+			this.say("nickserv", "identify " + pass);
+			irc_con.join(channel);
+		}).on("error", function(err) {
+			if (err.rawCommand !== "451") {
+				self.emit("error", err);
+			}
+		});
+	} else {
+		IRCConnections[host + ":" + server_port + "_" + name].join(channel);
+	}
+	var irc_con = IRCConnections[host + ":" + server_port + "_" + name];
+	
+	irc_con.on('names' + channel, function(nicks) {
+		users = Object.keys(nicks);
+		userRanks = nicks;
+	}).on('topic', function(chan, chan_topic, nick, message) {
+		if(channel === chan) {
+			self.emit("topic_changed", topic, chan_topic);
+			topic = chan_topic;
+		}
+	}).on('nick', function(old, newnick, channels, message) {
+		if (name === newnick) { return; }
+		self.emit("nick_changed", old, newnick, channels);
+	}).on('message' + channel, function (from, message) {
+		self.emit("irc_message", from, message);
+		if (message.startsWith(cmd_prefix)) {
+			var cmd_args = message.substring(0,1).split(" ");
+			var command = cmd_args.splice(0,1)[0].toLowerCase();
+			self.emit("command", command, from, cmd_args);
+			self.emit("command#" + command, from, cmd_args);
+		} else {
+			self.emit("message", from, message);
+		}
+	}).on('join' + channel, function(nick, message) {
+		if (name === nick) { return; }
+		irc_con.send("NAMES", channel);
+		irc_con.notice(nick, WelcomeMsg);
+		self.emit("join", nick);
+	}).on('kick' + channel, function(nick, by, reason, message) {
+		if (name === nick) { return; }
+		irc_con.send("NAMES", channel);
+		self.emit("kick", nick, by, reason);
+	}).on('part' + channel, function(nick, reason, message) {
+		if (name === nick) { return; }
+		irc_con.send("NAMES", channel);
+		self.emit("part", nick, reason);
+	}).on('quit', function(nick, reason, channels, message) {
+		if (name === nick) { return; }
+		irc_con.send("NAMES", channel);
+		self.emit("quit", nick, reason, channels);
+	}).on("kill", function(nick, reason, channels, message) {
+		if (name === nick) { return; }
+		irc_con.send("NAMES", channel);
+		self.emit("kill", nick, reason, channels);
+	});
+	
+	self.say = function(from, message) {
```

Comment:
Well, using explicit `return` statements would be a good start.  Choosing [prototypal inheritance **or** classical inheritance](http://stackoverflow.com/a/19640910) instead of mixing them would also help clarity.